File: core/src/com/xenoage/zong/core/music/format/Position.java
package com.xenoage.zong.core.music.format;

import lombok.Data;


/**
 * Custom positioning of an object.
 * 
 * @author Andreas Wenger
 */
@Data public final class Position
	implements Positioning {

	/** The x coordinate in mm, or null for default. */
	public final Float x;
	/** The y coordinate in LP, or null for default. */
	public final Float y;
	/** An additional horizontal offset in mm, or null for 0. */
	public final Float relativeX;
	/** An additional vertical offset in LP, or null for 0. */
	public final Float relativeY;
	

	/**
	 * Returns the additional horizontal offset of the given {@link Position},
	 * or 0 if the position is null or the offset is null.
	 */
	public static float getRelativeX(Position p) {
		return (p == null || p.relativeX == null ? 0 : p.relativeX);
	}


	/**
	 * Returns the additional vertical offset of the given {@link Position},
	 * or 0 if the position is null or the offset is null.
	 */
	public static float getRelativeY(Position p) {
		return (p == null || p.relativeY == null ? 0 : p.relativeY);
	}


	/**
	 * Gets the {@link Position} in the given {@link Positioning}, or null
	 * if it is no {@link Position}.
	 */
	public static Position asPosition(Positioning positioning) {
		if (positioning instanceof Position)
			return (Position) positioning;
		else
			return null;
	}


}

File: core/src/com/xenoage/zong/core/music/format/Position.kt
package com.xenoage.zong.core.music.format

import com.xenoage.utils.Mm
import com.xenoage.zong.core.music.LPf


/**
 * Custom positioning of an object.
 */
class Position(
	/** The x coordinate in mm, or null for default. */
	val x: Mm? = null,
	/** The y coordinate in [LPf], or null for default. */
	val y: LPf? = null,
	/** An additional horizontal offset in mm, or null for 0. */
	val relativeXOrNull: Mm? = null,
	/** An additional vertical offset in [LPf], or null for 0. */
	val relativeYOrNull: LPf? = null
) : Positioning {

	companion object {

		/**
		 * The additional horizontal offset of this [Position],
		 * or 0 if this position is null or the offset is null.
		 */
		val Position?.relativeX: Float
			get() = if (this == null || relativeX == null) 0f else relativeX

		/**
		 * The additional vertical offset of this [Position],
		 * or 0 if this position is null or the offset is null.
		 */
		val Position?.relativeY: Float
			get() = if (this == null || relativeY == null) 0f else relativeY
	}

}
